File: sdk/packages/typescript/src/pagination.ts
import type { Row } from './hydrate.js';

/** One page of hydrated rows. */
export interface Page {
  rows: Row[];
  totalCount: number | null;
  pageOffset: number;
}

/**
 * ListCall is the thenable returned by paginated facade methods: `await`
 * yields the first page's rows; `.pages()` iterates all pages.
 */
export class ListCall implements PromiseLike<Row[]> {
  private readonly fetchPage: (pageOffset: number) => Promise<Page>;
  private readonly pageSize: number;

  constructor(fetchPage: (pageOffset: number) => Promise<Page>, pageSize: number) {
    this.fetchPage = fetchPage;
    this.pageSize = pageSize;
  }

  then<TResult1 = Row[], TResult2 = never>(
    onfulfilled?: ((value: Row[]) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null,
  ): PromiseLike<TResult1 | TResult2> {
    return this.fetchPage(0).then(page => page.rows).then(onfulfilled, onrejected);
  }

  /** Iterates pages until a short page signals the end of the result set. */
  async *pages(): AsyncGenerator<Page> {
    let offset = 0;
    for (;;) {
      const page = await this.fetchPage(offset);
      yield page;
      if (page.rows.length < this.pageSize) return;
      offset += this.pageSize;
    }
  }

  /** Iterates individual rows across all pages. */
  async *[Symbol.asyncIterator](): AsyncGenerator<Row> {
    for await (const page of this.pages()) {
      for (const row of page.rows) yield row;
    }
  }
}

File: sdk/packages/python/src/whodb/_pagination.py
"""Pagination: list-shaped facade calls return ListCall objects."""

from __future__ import annotations

from collections.abc import Callable, Iterator
from typing import Any


class Page:
    """One page of hydrated rows."""

    def __init__(self, rows: list[dict], total_count: Any, page_offset: int):
        self.rows = rows
        self.total_count = total_count
        self.page_offset = page_offset


class ListCall:
    """Lazy list result: iterate rows, call .all() for the first page, or
    .pages() to walk every page (a short page terminates iteration)."""

    def __init__(self, fetch_page: Callable[[int], Page], page_size: int):
        self._fetch_page = fetch_page
        self._page_size = page_size

    def all(self) -> list[dict]:
        """Return the first page's rows (mirrors awaiting the TS ListCall)."""
        return self._fetch_page(0).rows

    def pages(self) -> Iterator[Page]:
        """Iterate pages until a short page signals the end."""
        offset = 0
        while True:
            page = self._fetch_page(offset)
            yield page
            if len(page.rows) < self._page_size:
                return
            offset += self._page_size

    def __iter__(self) -> Iterator[dict]:
        for page in self.pages():
            yield from page.rows
